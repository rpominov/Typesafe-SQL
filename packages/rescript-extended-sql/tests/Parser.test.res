open Jest

each(
  [
    "SELECT 1",
    "-- test\nSELECT -- test2\n--\n 1-- test3",
    "/* test \n abc*/\nSELECT /* test2 /* test3 */ */ 1/**/",
    "SELECT 1/* test",
    "SELECT 1/* test /* test2 ",
    "/*@name: test*/SELECT 1",
    "/*\n * @name: test\n */SELECT 1",
    "/*@name: testA4_\n@name: test1*/SELECT 1",
    "/*@name: %abc*/SELECT 1",
    "/*@name: 4abc*/SELECT 1",
    "/*@name: */SELECT 1",
    "/*abc @name: abc*/SELECT 1",
    "-- abc\n/* @name: abc*/SELECT 1",
    "SELECT 1/* @name: abc*/",
    "SELECT :foo = :bar",
    "SELECT ':'",
    "SELECT '\\:'",
    "SELECT :num:raw<1|2>",
    "SELECT :num:raw<>",
    "SELECT '\\:num\\:raw<1|2>'",
    "SELECT :num:raw<1,|> 2",
    "SELECT :num:raw<<<1<|>|||2>>>",
    "SELECT :num:raw<<<1|||2>>",
    "INSERT INTO test (foo, bar) VALUES :values:batch<(:foo, :bar)>",
    "INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo:batch<:bar>)>>",
    "INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo:raw<foo|bar>)>>",
    "INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo, :bar)>",
    "INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo /* <comment> */)>>",
  ],
  "parse(\"%s\")",
  code => code->Parser.parse->expect->toMatchSnapshot,
)

// TODO: more tests for parseFile
each(["SELECT 1;SELECT 2;", "-- @separator:### \nSELECT 1###SELECT 2"], "parseFile(\"%s\")", code =>
  code->Parser.parseFile->expect->toMatchSnapshot
)
