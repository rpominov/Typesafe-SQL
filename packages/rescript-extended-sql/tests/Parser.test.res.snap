// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Batch (nested batch) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 34,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "INSERT INTO test (foo, bar) VALUES ",
        },
      },
      Object {
        "end": 69,
        "start": 35,
        "val": Object {
          "TAG": 5,
          "_0": "values",
          "_1": ",",
          "_2": Array [
            Object {
              "end": 50,
              "start": 50,
              "val": Object {
                "TAG": 0,
                "_0": "(",
              },
            },
            Object {
              "end": 66,
              "start": 51,
              "val": Object {
                "TAG": 5,
                "_0": "foo",
                "_1": ",",
                "_2": Array [
                  Object {
                    "end": 65,
                    "start": 62,
                    "val": Object {
                      "TAG": 3,
                      "_0": "bar",
                    },
                  },
                ],
              },
            },
            Object {
              "end": 67,
              "start": 67,
              "val": Object {
                "TAG": 0,
                "_0": ")",
              },
            },
          ],
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Batch (nested comment) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 34,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "INSERT INTO test (foo, bar) VALUES ",
        },
      },
      Object {
        "end": 73,
        "start": 35,
        "val": Object {
          "TAG": 5,
          "_0": "values",
          "_1": ",",
          "_2": Array [
            Object {
              "end": 50,
              "start": 50,
              "val": Object {
                "TAG": 0,
                "_0": "(",
              },
            },
            Object {
              "end": 54,
              "start": 51,
              "val": Object {
                "TAG": 3,
                "_0": "foo",
              },
            },
            Object {
              "end": 55,
              "start": 55,
              "val": Object {
                "TAG": 0,
                "_0": " ",
              },
            },
            Object {
              "end": 70,
              "start": 56,
              "val": Object {
                "TAG": 2,
                "_0": " <comment> ",
              },
            },
            Object {
              "end": 71,
              "start": 71,
              "val": Object {
                "TAG": 0,
                "_0": ")",
              },
            },
          ],
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Batch (nested raw) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 34,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "INSERT INTO test (foo, bar) VALUES ",
        },
      },
      Object {
        "end": 70,
        "start": 35,
        "val": Object {
          "TAG": 5,
          "_0": "values",
          "_1": ",",
          "_2": Array [
            Object {
              "end": 50,
              "start": 50,
              "val": Object {
                "TAG": 0,
                "_0": "(",
              },
            },
            Object {
              "end": 67,
              "start": 51,
              "val": Object {
                "TAG": 4,
                "_0": "foo",
                "_1": Array [
                  "foo",
                  "bar",
                ],
              },
            },
            Object {
              "end": 68,
              "start": 68,
              "val": Object {
                "TAG": 0,
                "_0": ")",
              },
            },
          ],
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Batch (not closed) 1`] = `
Object {
  "TAG": 1,
  "_0": Object {
    "end": undefined,
    "start": 63,
    "val": "Was expecting a batch parameter close sequence >>, but reached the end of the string",
  },
}
`;

exports[`Batch 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 34,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "INSERT INTO test (foo, bar) VALUES ",
        },
      },
      Object {
        "end": 61,
        "start": 35,
        "val": Object {
          "TAG": 5,
          "_0": "values",
          "_1": ",",
          "_2": Array [
            Object {
              "end": 49,
              "start": 49,
              "val": Object {
                "TAG": 0,
                "_0": "(",
              },
            },
            Object {
              "end": 53,
              "start": 50,
              "val": Object {
                "TAG": 3,
                "_0": "foo",
              },
            },
            Object {
              "end": 55,
              "start": 54,
              "val": Object {
                "TAG": 0,
                "_0": ", ",
              },
            },
            Object {
              "end": 59,
              "start": 56,
              "val": Object {
                "TAG": 3,
                "_0": "bar",
              },
            },
            Object {
              "end": 60,
              "start": 60,
              "val": Object {
                "TAG": 0,
                "_0": ")",
              },
            },
          ],
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Block comments 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 14,
        "start": 0,
        "val": Object {
          "TAG": 2,
          "_0": " test 
 abc",
        },
      },
      Object {
        "end": 22,
        "start": 15,
        "val": Object {
          "TAG": 0,
          "_0": "
SELECT ",
        },
      },
      Object {
        "end": 45,
        "start": 23,
        "val": Object {
          "TAG": 2,
          "_0": " test2 /* test3 */ ",
        },
      },
      Object {
        "end": 47,
        "start": 46,
        "val": Object {
          "TAG": 0,
          "_0": " 1",
        },
      },
      Object {
        "end": 51,
        "start": 48,
        "val": Object {
          "TAG": 2,
          "_0": "",
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Block comments error 1`] = `
Object {
  "TAG": 1,
  "_0": Object {
    "end": undefined,
    "start": 15,
    "val": "Was expecting a block comment close sequence */, but reached the end of the string",
  },
}
`;

exports[`Block comments error nested 1`] = `
Object {
  "TAG": 1,
  "_0": Object {
    "end": undefined,
    "start": 25,
    "val": "Was expecting a block comment close sequence */, but reached the end of the string",
  },
}
`;

exports[`Inline comments 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 6,
        "start": 0,
        "val": Object {
          "TAG": 1,
          "_0": " test",
        },
      },
      Object {
        "end": 14,
        "start": 7,
        "val": Object {
          "TAG": 0,
          "_0": "
SELECT ",
        },
      },
      Object {
        "end": 22,
        "start": 15,
        "val": Object {
          "TAG": 1,
          "_0": " test2",
        },
      },
      Object {
        "end": 23,
        "start": 23,
        "val": Object {
          "TAG": 0,
          "_0": "
",
        },
      },
      Object {
        "end": 25,
        "start": 24,
        "val": Object {
          "TAG": 1,
          "_0": "",
        },
      },
      Object {
        "end": 28,
        "start": 26,
        "val": Object {
          "TAG": 0,
          "_0": "
 1",
        },
      },
      Object {
        "end": 36,
        "start": 29,
        "val": Object {
          "TAG": 1,
          "_0": " test3",
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Name attribute (2nd ignored) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 30,
        "start": 0,
        "val": Object {
          "TAG": 2,
          "_0": "@name: testA4_
@name: test1",
        },
      },
      Object {
        "end": 38,
        "start": 31,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT 1",
        },
      },
    ],
    "attributes": Object {
      "name": "testA4_",
    },
  },
}
`;

exports[`Name attribute (after code) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 7,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT 1",
        },
      },
      Object {
        "end": 22,
        "start": 8,
        "val": Object {
          "TAG": 2,
          "_0": " @name: abc",
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Name attribute (empty) 1`] = `
Object {
  "TAG": 1,
  "_0": Object {
    "end": 10,
    "start": 0,
    "val": "Invalid @name attribute: ",
  },
}
`;

exports[`Name attribute (invalid) 1`] = `
Object {
  "TAG": 1,
  "_0": Object {
    "end": 14,
    "start": 0,
    "val": "Invalid @name attribute: %abc",
  },
}
`;

exports[`Name attribute (invalid, starts with a digit) 1`] = `
Object {
  "TAG": 1,
  "_0": Object {
    "end": 14,
    "start": 0,
    "val": "Invalid @name attribute: 4abc",
  },
}
`;

exports[`Name attribute (many comments) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 5,
        "start": 0,
        "val": Object {
          "TAG": 1,
          "_0": " abc",
        },
      },
      Object {
        "end": 6,
        "start": 6,
        "val": Object {
          "TAG": 0,
          "_0": "
",
        },
      },
      Object {
        "end": 21,
        "start": 7,
        "val": Object {
          "TAG": 2,
          "_0": " @name: abc",
        },
      },
      Object {
        "end": 29,
        "start": 22,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT 1",
        },
      },
    ],
    "attributes": Object {
      "name": "abc",
    },
  },
}
`;

exports[`Name attribute (text before) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 17,
        "start": 0,
        "val": Object {
          "TAG": 2,
          "_0": "abc @name: abc",
        },
      },
      Object {
        "end": 25,
        "start": 18,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT 1",
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Name attribute 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 14,
        "start": 0,
        "val": Object {
          "TAG": 2,
          "_0": "@name: test",
        },
      },
      Object {
        "end": 22,
        "start": 15,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT 1",
        },
      },
    ],
    "attributes": Object {
      "name": "test",
    },
  },
}
`;

exports[`Parameters (no name, error) 1`] = `
Object {
  "TAG": 1,
  "_0": Object {
    "end": 9,
    "start": 8,
    "val": "Unexpected : symbol not followed by a parameter name. If you meant to simply insert :, please escape it with a backslash \\\\:",
  },
}
`;

exports[`Parameters (no name, escaped) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 10,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT ':'",
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Parameters 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 6,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT ",
        },
      },
      Object {
        "end": 10,
        "start": 7,
        "val": Object {
          "TAG": 3,
          "_0": "foo",
        },
      },
      Object {
        "end": 13,
        "start": 11,
        "val": Object {
          "TAG": 0,
          "_0": " = ",
        },
      },
      Object {
        "end": 17,
        "start": 14,
        "val": Object {
          "TAG": 3,
          "_0": "bar",
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Parse file (custom separator) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "separator": Array [
      "#",
      "#",
      "#",
    ],
    "statements": Array [
      Object {
        "ast": Array [
          Object {
            "end": 26,
            "start": 18,
            "val": Object {
              "TAG": 0,
              "_0": "
SELECT 1",
            },
          },
        ],
        "attributes": Object {
          "name": undefined,
        },
      },
      Object {
        "ast": Array [
          Object {
            "end": 37,
            "start": 30,
            "val": Object {
              "TAG": 0,
              "_0": "SELECT 2",
            },
          },
        ],
        "attributes": Object {
          "name": undefined,
        },
      },
    ],
  },
}
`;

exports[`Parse file 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "separator": Array [
      ";",
    ],
    "statements": Array [
      Object {
        "ast": Array [
          Object {
            "end": 7,
            "start": 0,
            "val": Object {
              "TAG": 0,
              "_0": "SELECT 1",
            },
          },
        ],
        "attributes": Object {
          "name": undefined,
        },
      },
      Object {
        "ast": Array [
          Object {
            "end": 16,
            "start": 9,
            "val": Object {
              "TAG": 0,
              "_0": "SELECT 2",
            },
          },
        ],
        "attributes": Object {
          "name": undefined,
        },
      },
    ],
  },
}
`;

exports[`Raw (<<<) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 6,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT ",
        },
      },
      Object {
        "end": 28,
        "start": 7,
        "val": Object {
          "TAG": 4,
          "_0": "num",
          "_1": Array [
            "1<|>",
            "2",
          ],
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Raw (<<<, not closed) 1`] = `
Object {
  "TAG": 1,
  "_0": Object {
    "end": undefined,
    "start": 25,
    "val": "Was expecting a raw parameter close sequence >>>, but reached the end of the string",
  },
}
`;

exports[`Raw (empty option) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 6,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT ",
        },
      },
      Object {
        "end": 19,
        "start": 7,
        "val": Object {
          "TAG": 4,
          "_0": "num",
          "_1": Array [
            "1,",
            "",
          ],
        },
      },
      Object {
        "end": 21,
        "start": 20,
        "val": Object {
          "TAG": 0,
          "_0": " 2",
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Raw (escaped) 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 23,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT ':num:raw<1|2>'",
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Raw 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 6,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT ",
        },
      },
      Object {
        "end": 19,
        "start": 7,
        "val": Object {
          "TAG": 4,
          "_0": "num",
          "_1": Array [
            "1",
            "2",
          ],
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;

exports[`Simplest query 1`] = `
Object {
  "TAG": 0,
  "_0": Object {
    "ast": Array [
      Object {
        "end": 7,
        "start": 0,
        "val": Object {
          "TAG": 0,
          "_0": "SELECT 1",
        },
      },
    ],
    "attributes": Object {
      "name": undefined,
    },
  },
}
`;
